doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,247 @@
+        driven triggering.
+
+  - add class `ConditionVariableWithOriginTracingSignaler` (I know, better name)
+  
+    **Reason:** it behaves quiete differently then the `ConditionVariable`. The `ConditionVariable` tells you:
+    someone has signalled you and you have to find out who it was. The `ConditionVariableWithOriginTracing`
+    would tell you: A, B and C have signalled you.
+
+    ```cpp
+    class ConditionVariableWithOriginTracingSignaler {
+      public:
+        // originId = corresponds to the id in the ConditionVariable atomic_bool array to 
+        // identify the trigger origin easily
+        void notify(const uint64_t originId); 
+
+        // alternative: origin is set in the ctor and we just call notify
+        ConditionVariableWithOriginTracingSignaler(const uint64_t originId);
+        void notify();
+    };
+    ```
+
+  - add class `ConditionVariableWithOriginTracingWaiter`
+  ```cpp
+  class ConditionVariableWithOriginTracingWaiter {
+    public:
+      // returns a list of ids which have notified the conditionVariable
+      cxx::vector<uint64_t> wait(); 
+      // sets the atomic bool to false in m_triggeredBy[id]
+      void resetWaitState(const uint64_t id) noexcept; 
+  };
+  ```
+
+#### ReactAl 
+```cpp
+enum class ReactAlErrors {
+  CALLBACK_CAPACITY_EXCEEDED
+};
+
+template<uint64_t CallbackCapacity>
+class ReactAl {
+  public:
+    template<typename EventOrigin, typename EventType>
+    cxx::expected<ReactAlErrors> attachEvent(
+      EventOrigin & origin, const EventType & eventType,
+      const cxx::function_ref<void(EventOrigin&)> & callback
+    ) noexcept;
+
+    template<typename EventOrigin, typename EventType>
+    void detachEvent( EventOrigin & origin, const EventType & eventType );
+};
+
+```
+ 
+  - Contains a thread which will wake up and iterate through ConditionVariable
+    bool array. If an entry is true, set it to false and then call the 
+    corresponding callback.
+
+#### Class which is attachable to ReactAl 
+
+```cpp
+class SomeSubscriber {
+  //...
+  // index is the index to the bool inside of the condition variable, maybe it 
+  // is later hidden inside of some abstraction
+  //
+  // OutOfScopeCallback_t - callback which should be called when SomeSubscriber
+  //                        goes out of scope and has to deregister itself from 
+  //                        the reactal
+  friend class ReactAl;
+  private:
+    void attachCallback(SubscriberEvent event, ConditionVariable & cv, ConditionVariable::index_t & index
+                        const OutOfScopeCallback_t & callback);
+
+    void detachCallback(SubscriberEvent event);
+  //...
+};
+```
+
+The `ReactAl` will call the methods above to attach or detach a condition 
+variable to a class. The methods should be private and the `ReactAl` must then 
+be declared as a friend.
+
+## Open Points
+
+ - Maybe introduce some abstraction Condition Variable tracing handling. 
```

Comment:
I would try to generalize the trigger mechanism to accept very general objects which just need to provide a wait/notify mechanism. (As a template parameter Notifyable, most likely). There maybe the necessity to also rely on a factory method ``Notifyable::create(Args...&& args)`` or something similar for construction in specific cases.

Reply:
Here I would say again YAGNI until the use case arises. To create a much more general approach would be nice but it also has some overhead since we have to create the ReactAL then with multiple template parameters, have to provide factory methods etc..